docs/proposals/201812_thanos-remote-receive.md

```diff
@@ -0,0 +1,174 @@
+
+This component is only recommended for uses, where pushing is the only viable solution, for example analytics use cases or cases where the data ingestion must be client initiated, for example a software as a service type environments.
+
+This component is not recommended in order to achieve a global view of data of a single tenant, for those cases the sidecar based approach with layered Thanos queriers is recommended. Also users are asked to note the [various pros and cons of pushing metrics](https://docs.google.com/document/d/1H47v7WfyKkSLMrR8_iku6u9VB73WrVzBHb2SB6dL9_g/edit#heading=h.2v27snv0lsur).
+
+## Technical summary of Thanos receiver component
+
+To solve the above mentioned problems, a new Thanos component is proposed: the Thanos receiver.
+
+Prometheus has the remote write API to send samples collected by a Prometheus server to a remote location. This is essentially replicating the Prometheus write-ahead-log to that remote location. The Thanos receiver represents the remote location that accepts the Prometheus remote write API. This brings some interesting characteristics with it. It retains the metadata aspects that has contributed to a large part to Prometheus' success, while the remote location can add additional features to the stack that are not present (and arguably shouldn't be) in Prometheus, such as long term storage and horizontal scalability, while keeping the local Prometheus useful for reliable alerting on short term data.
+
+## Architecture
+
+The Thanos receiver component seamlessly integrates into the rest of the Thanos components. It acts similarly to what is referred to in Thanos as a "source", in the current set of components, this is typically represented by the Thanos sidecar that is put next to Prometheus to ship tsdb blocks into object storage and reply to store API requests, however in the case of the Thanos receiver, the Thanos sidecar is not necessary anymore, as the data is replicated from the original Prometheus server to the Thanos receiver, and the Thanos receiver participates in the Thanos gossip mesh. The Prometheus server on a tenant’s infrastructure can therefore be completely vanilla and is just configured to replicate its time-series to the Thanos receiver.
+
+Instead of directly scraping metrics, however, the Thanos receiver accepts Prometheus remote write requests, and writes these into a local instance of the Prometheus tsdb, to prevent data leaking at the database level, each tenant has an individual tsdb instance, meaning a single Thanos receiver may manage multiple tsdb instances. When the receiver is queried, the respective tsdb is used to retrieve data. Furthermore, like the Thanos sidecar, the receiver answers Thanos store API requests and uploads built blocks of the Prometheus tsdb. Implementation-wise, this just requires wiring up existing components.
```

Comment:
Can we expand this? 
` to prevent data leaking at the database level, each tenant has an individual tsdb instance, meaning a single Thanos receiver may manage multiple tsdb instances.`

So:

Multiple of receivers will put same tenant into multiple of TSDB instances, resulting in multiple blocks per single tenant, right? How their external labels would looks like, as they are required to be unique when stored in single object storage (otherwise, how compactor can know which compact with which?). [Vertical compaction](https://github.com/prometheus/tsdb/pull/370) might help to save data size, and number of different blocks to lookup.

If my assumptions here are correct, do we have rough idea how many different blocks for single tenant we can have in system? Do we have limit? For large number of tenants, when you want large number of receivers I can imagine this being quite large.. e.g tenant with 1k series and  20 blocks for those ;p

Reply:
hmm.. I may very well be missing something, but why must they be unique? As I understand the vertical compaction (which, yes it is indeed required, I'll add it to the end of the document) it relies on tsdbs with the same identity and overlapping time.

The more I think about it, the more I feel we may need a second layer of sharding to determine which pool of soft tenants a tenant belongs to, that way we can limit the number of tenants per receiver node rather easily.

Reply:
Added an additional `hashmod` selection mechanism for hashring selection.

Reply:
> hmm.. I may very well be missing something, but why must they be unique? As I understand the vertical compaction (which, yes it is indeed required, I'll add it to the end of the document) it relies on tsdbs with the same identity and overlapping time.

This is tricky, because:
- vertical compaction is not used anywhere in production.
- there is no plan to support vertical compaction out of box for Thanos, because there was no use case of  ( until this proposal (: ). 
- because buckets are shared, it is easy to misconfigure and merge vertically blocks that you shouldn't.. (misconfigured external labels)

But yea, seems like it is must have for this, so worth to put it as depedencendy of it (if it's not already) (: 

Making it more efficient (less vs more blocks that are required to be merged vertically) is separate  issue I guess, we need to data to decide on this. But having vertical compact in first place is must have ;p